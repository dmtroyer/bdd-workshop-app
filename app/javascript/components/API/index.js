import axios from 'axios';

const fetchUserProfile = async id => {
  const response = await axios
    .get(`/api/v1/profiles/${id}`, {
      // TODO: set derault for all calls
      headers: { Accept: 'application/json' },
    })
    .catch(() => {
      return createUserProfile();
    });
  return response;
};

const createUserProfile = async () => {
  const response = await axios.post(
    '/api/v1/profiles',
    {},
    {
      headers: { Accept: 'application/json' },
    }
  );
  return response;
};

const updateUserProfile = async ({ data: { handle, id } }) => {
  const response = await axios.put(
    `/api/v1/profiles/${id}`,
    { player: { handle } },
    {
      headers: { Accept: 'application/json' },
    }
  );
  return response;
};

export { fetchUserProfile };
export { createUserProfile };
export { updateUserProfile };
export default { fetchUserProfile, createUserProfile, updateUserProfile };
