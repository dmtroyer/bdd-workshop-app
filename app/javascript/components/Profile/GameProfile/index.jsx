import React from 'react';
import { Button } from 'reactstrap';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

const GameProfile = ({ profile }) => {
  return (
    <>
      {profile.percentComplete === 100 ? (
        <p>Your profile is complete!</p>
      ) : (
        <Button
          data-testid="actions"
          tag={Link}
          to="/profile"
          className="float-right"
          color="primary"
        >
          Complete my profile
        </Button>
      )}
    </>
  );
};

GameProfile.propTypes = {
  profile: PropTypes.object.isRequired,
};

export default GameProfile;
