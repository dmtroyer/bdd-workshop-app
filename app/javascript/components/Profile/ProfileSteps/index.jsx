import React from 'react';
import Avatar from '../../Avatar';
import ProgressBar from '../../ProgressBar';
import { Button } from 'reactstrap';
import { Link } from 'react-router-dom';
import OurForm from '../../OurForm';
import PropTypes from 'prop-types';

const ProfileSteps = ({ profile, onSubmit }) => (
  <div>
    <h1>
      Hi : <span data-testid="details-handle">{profile.handle}</span>
    </h1>
    <p>Your profile is almost complete</p>
    <ProgressBar percentComplete={profile.percentComplete} />
    <Avatar profile={profile} />
    <OurForm onSubmit={onSubmit} profile={profile} />
    <div data-testid="actions">
      <Button tag={Link} to="/game" color="secondary">
        Play the game
      </Button>
    </div>
  </div>
);

ProfileSteps.propTypes = {
  profile: PropTypes.object.isRequired,
  onSubmit: PropTypes.func.isRequired,
};

export default ProfileSteps;
