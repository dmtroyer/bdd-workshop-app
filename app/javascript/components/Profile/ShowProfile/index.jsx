import React from 'react';
import { Row, Col } from 'reactstrap';
import PropTypes from 'prop-types';
import Avatar from '../../Avatar';

const ShowProfile = ({ profile }) => {
  return (
    <>
      <Avatar profile={profile} />
      <Row>
        <Col>
          <dl>
            <dt>handle</dt>
            <dd data-testid="details-handle">{profile.handle}</dd>
            <dt>email</dt>
            <dd data-testid="details-email">{profile.email}</dd>
          </dl>
        </Col>
      </Row>
    </>
  );
};
ShowProfile.propTypes = {
  profile: PropTypes.shape({
    id: PropTypes.string.isRequired,
    handle: PropTypes.string,
    email: PropTypes.string,
  }),
};

export default ShowProfile;
