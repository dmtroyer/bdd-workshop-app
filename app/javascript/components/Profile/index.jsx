import React from 'react';
import { Spinner } from 'reactstrap';

function Profile() {
  return (
    <div data-test-id="profile">
      <h1>Profile</h1>
      <Spinner color="primary" data-test-id="loading-profile" />
    </div>
  );
}

export default Profile;
