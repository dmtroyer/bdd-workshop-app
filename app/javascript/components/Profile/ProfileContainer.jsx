import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import API from '../API';
import MainContainer from '../MainContainer';
import messageBus from '../../utils/messageBus';
import PropTypes from 'prop-types';
import ShowProfile from './ShowProfile';
import ProfileSteps from './ProfileSteps';
import OurForm from '../OurForm';

class ProfileContainer extends Component {
  constructor(props) {
    super(props);
  }

  state = {
    isEditting: false,
  };

  toggleIsEditting() {
    this.setState({ isEditting: !this.state.isEditting });
  }

  updateUserProfile(data) {
    data = Object.assign(this.props.profile, data);
    return API.updateUserProfile({ data })
      .then(response => {
        messageBus.info('Updated user profile');
        this.props.fetchProfile();
        if (this.state.isEditting) {
          this.toggleIsEditting();
        }
        return Promise.resolve(response);
      })
      .catch(res => {
        const errors = res.response.data.errors;
        const messages = Array.isArray(errors)
          ? errors.map(e => e.detail).join()
          : Object.keys(errors)
              .map(key => [key, errors[key]].join(': '))
              .join();
        messageBus.error(messages);
      });
  }

  render() {
    const { isEditting } = this.state;
    const { profile } = this.props;
    if (!profile) {
      return <Redirect to="/" />;
    }
    return (
      <MainContainer dataTestId="profile">
        {isEditting && (
          <OurForm
            onSubmit={this.updateUserProfile.bind(this)}
            profile={profile}
          />
        )}
        {profile.percentComplete === 100 && !isEditting && (
          <ShowProfile
            profile={profile}
            editOnClick={this.toggleIsEditting.bind(this)}
          />
        )}
        {profile.percentComplete !== 100 && (
          <ProfileSteps
            profile={profile}
            onSubmit={this.updateUserProfile.bind(this)}
          />
        )}
      </MainContainer>
    );
  }
}

ProfileContainer.propTypes = {
  profile: PropTypes.object,
  fetchProfile: PropTypes.func,
  history: PropTypes.object,
};

export default ProfileContainer;
