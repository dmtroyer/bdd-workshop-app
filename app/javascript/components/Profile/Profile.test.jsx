import React from 'react';
import { shallow } from 'enzyme';
import Profile from '.';
import API from '../API';

describe('Profile', () => {
  it('renders correctly in loading state', () => {
    const wrapper = shallow(<Profile />);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles invalid player being set', () => {
    localStorage.setItem('player', undefined);
    API.createUserProfile = () =>
      Promise.resolve({
        response: { profile: 'the-id' },
      });

    const wrapper = shallow(<Profile />);

    return wrapper.instance().userProfilePromise.then(() => {
      // TODO currently rendering spinner
      expect(wrapper).toMatchSnapshot();
      // TODO test for the-id to be output
      // TODO test the id has been set to localstorage
    });
  });

  it('renders profile as returned from fetch the-id', () => {
    localStorage.setItem('player', JSON.stringify({ id: 'the-id' }));
    API.fetchUserProfile = id =>
      Promise.resolve({
        response: { profile: id },
      });

    const wrapper = shallow(<Profile />);

    return wrapper.instance().userProfilePromise.then(() => {
      // TODO currently rendering spinner
      expect(wrapper).toMatchSnapshot();
      // TODO test for the-id to be output
    });
  });

  it('renders profile as created as no player found', () => {
    localStorage.setItem('player', null);
    API.createUserProfile = id =>
      Promise.resolve({
        response: { profile: id },
      });

    const wrapper = shallow(<Profile />);

    return wrapper.instance().userProfilePromise.then(() => {
      // TODO currently rendering spinner
      expect(wrapper).toMatchSnapshot();
      // TODO test for the-id to be output
    });
  });

  // TODO: sort out what should be displayed when the profile fetch fails
  // eslint-disable-next-line jest/no-disabled-tests
  it.skip('renders an error if profile fetch fails from localstorage the-id', () => {
    localStorage.setItem('player', JSON.stringify({ id: 'the-id' }));
    API.fetchUserProfile = id =>
      Promise.reject({
        response: {
          status: 500,
          statusText: `catastrophic failure in fetch ${id}`,
        },
      });
    // TODO should it not go onto createUserProfile?
    // API.createUserProfile = () =>
    //   Promise.reject({
    //     response: { status: 500, statusText: 'catastrophic failure in create' },
    //   });

    const wrapper = shallow(<Profile />);

    return wrapper.instance().userProfilePromise.then(() => {
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('.alert').prop('errorMessage')).toEqual(
        '500 - catastrophic failure in fetch the-id'
      );
    });
  });
});
