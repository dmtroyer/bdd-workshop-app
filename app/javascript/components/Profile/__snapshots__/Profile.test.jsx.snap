// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile renders correctly 1`] = `
<div
  data-test-id="profile"
>
  <h1>
    Profile
  </h1>
  <Spinner
    color="primary"
    data-test-id="loading-profile"
    tag="div"
    type="border"
  >
    Loading...
  </Spinner>
</div>
`;
